src/Eshopworld.Messaging/ServiceBusAdapter.cs

```diff
@@ -121,15 +135,23 @@ internal async Task Lock(T message)
         /// <param name="_">[Ignored]</param>
         internal async Task Read([CanBeNull]object _)
         {
+            if (Receiver.IsClosedOrClosing) return;
+
             var messages = await Receiver.ReceiveAsync(BatchSize).ConfigureAwait(false);
             if (messages == null) return;
 
             foreach (var message in messages)
             {
-                var messageBody = JsonConvert.DeserializeObject<T>(Encoding.UTF8.GetString(message.Body));
-
-                Messages[messageBody] = message;
-                MessagesIn.OnNext(messageBody);
+                if (RawMessages)
+                {
+                    MessagesIn.OnNext(message as T);
+                }
+                else
+                {
+                    var messageBody = JsonConvert.DeserializeObject<T>(Encoding.UTF8.GetString(message.Body));
+                    Messages[messageBody] = message;

```

Comment:
The only problem I see here is there are multi-messages in the dictionary and they have the exact same body then the first insert may be overwritten. But I'm thinking the assumption is that's ok given it's a message per type. But can there be scenarios where the headers of the msg are the distinguishing element and are used in code/business logic?